core/sdk-core/src/main/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -87,6 +78,20 @@ public URI get(String accessKey, EndpointDiscoveryRequest request) {
         return endpoint.endpoint();
     }
 
+    public CompletableFuture<URI> getAsync(String accessKey, EndpointDiscoveryRequest request) {
+        String key = getKey(accessKey, request);
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        // If a service call needs to be made to discover endpoint
+        // a completable future for the service call is returned, unblocking I/O
+        // and then completed asynchronously
+        if (endpoint == null && request.required()) {
+            return discoverEndpointHandler(key, request);
+        }
+        // In the event of a cache hit, i.e. service call not required, defer to the synchronous code path method.
+        return CompletableFuture.completedFuture(get(accessKey, request));
```

Comment:
If we call get here, it'll go through computing key and checking cache again. Can we extract the part where we refresh cache to a separate method and just call that here?

Reply:
good callout. I will do that